cmd/output/output.go

```diff
@@ -15,31 +15,98 @@
 
-func ParseOutputFormat(outputFormat string) (*OutputFormat, error) {
-	format := new(OutputFormat)
+// ParseOutputFormat returns an OutputFormater based on the specified format.
+// Accepted values include 'json', 'yaml', and 'jsonpath=PATH'.
+// Empty string defaults to 'json'.
+// For more about JSONPath, see https://goessner.net/articles/JsonPath/
+func ParseOutputFormat(outputFormat string) (OutputFormater, error) {
 	switch {
-	case outputFormat == "":
-		return format, nil
-	case strings.HasPrefix(outputFormat, "jsonpath="):
+	case outputFormat == "" || outputFormat == "json":
+		return MarshalToJson, nil
+	case outputFormat == "yaml":
+		return MarshalToYaml, nil
+	case strings.HasPrefix(outputFormat, "jsonpath=") && outputFormat != "jsonpath=":
 		toks := strings.Split(outputFormat, "=")
 		if len(toks) != 2 {
 			return nil, errors.New(fmt.Sprintf("Failed to parse output format flag value: %s", outputFormat))
 		}
-		format.JsonPath = toks[1]
-		break
+		return MarshalToJsonPathWrapper(toks[1]), nil
 	default:
 		return nil, errors.New(fmt.Sprintf("Failed to parse output format flag value: %s", outputFormat))
 	}
-	return format, nil
+}
+
+func MarshalToJson(input interface{}) ([]byte, error) {
+	pretty, err := json.MarshalIndent(input, "", " ")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal to json: %v", err)
+	}
+	return pretty, nil
+}
+
+// MarshalToJsonPathWrapper returns a MarshalToJsonPath function that uses the
+// specified jsonpath expression.
+// This leverages the kubernetes jsonpath library
+// (https://kubernetes.io/docs/reference/kubectl/jsonpath/).
+func MarshalToJsonPathWrapper(expression string) OutputFormater {
+	expr := expression
+	// aka MarshalToJsonPath
+	return func(input interface{}) ([]byte, error) {
+		jp := jsonpath.New("json-path")
+		buffer := new(bytes.Buffer)
+
+		if err := jp.Parse(expr); err != nil {
+			return buffer.Bytes(), fmt.Errorf("Failed to parse jsonpath expression: %v", err)
+		}
+
+		err := jp.Execute(buffer, input)
+		if err != nil {
+			return buffer.Bytes(), fmt.Errorf("Failed to execute jsonpath %s on input %s: %v ", expr, input, err)
+		}
+
+		// unquote since go quotes the string if the bytes is a string.
+		return unquote(buffer.Bytes()), nil
+	}
+}
+
+func MarshalToYaml(input interface{}) ([]byte, error) {
+	pretty, err := yaml.Marshal(input)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal to yaml: %v", err)
+	}
+	return pretty, nil
+}
+
+// TODO: is this nessesary when using bytes.Bytes() instead of bytes.String() ?
+func unquote(input []byte) []byte {
+	input = bytes.TrimLeft(input, "\"")
+	input = bytes.TrimRight(input, "\"")
+	return input
+}
+
+// parseJsonPath finds the values specified in the input data as specified with the template.
+// This leverages the kubernetes jsonpath libs (https://kubernetes.io/docs/reference/kubectl/jsonpath/).
+func parseJsonPath(input interface{}, template string) (*bytes.Buffer, error) {
```

Comment:
looks like this is unused... probably got refactored into the closure inside `MarshalToJsonPathWrapper`?

Reply:
yup. will delete.